File: experiments/enumerate_state_space.py
"""Exhaustively enumerate reachable positional states of Ayo Olopon.

This experiment intentionally uses positional transposition keys only:
history, including ``_positions_since_capture``, is not part of the key.
The game implementation remains the sole source of legal actions and state
transitions.
"""

from __future__ import annotations

import argparse
import json
import pickle
import sys
import time
import tracemalloc
from collections import Counter
from pathlib import Path
from typing import Any

import pyspiel

# Make the repository root importable when this file is run directly from the
# experiments directory (``python experiments/enumerate_state_space.py``).
REPOSITORY_ROOT = Path(__file__).resolve().parents[1]
if str(REPOSITORY_ROOT) not in sys.path:
    sys.path.insert(0, str(REPOSITORY_ROOT))

# Importing the package registers the game with OpenSpiel.
from Model.ayo_olopon import ayo_olopon  # noqa: F401


TOTAL_SEEDS = 48
EXPECTED_HOUSES = 12
DEFAULT_OUTPUT = Path("experiments/ayo_olopon_state_space.json")
DEFAULT_CHECKPOINT = Path("experiments/ayo_olopon_state_space.checkpoint.pkl")


class EnumerationInvariantError(RuntimeError):
    """Raised when a generated state violates an experiment invariant."""


def canonical_state(state: pyspiel.State) -> tuple[tuple[int, ...], tuple[int, ...], int]:
    """Return exactly (board, captured, player) as an immutable key."""
    return (
        tuple(int(value) for value in state.board),
        tuple(int(value) for value in state.captured),
        int(state.current_player()),
    )


def validate_state(state: pyspiel.State, key: tuple[Any, ...], context: str) -> None:
    """Fail loudly if the model produces an invalid positional state."""
    board, captured, player = key
    if len(board) != EXPECTED_HOUSES:
        raise EnumerationInvariantError(
            f"{context}: expected 12 houses, got {len(board)}; state={key}"
        )
    if any(not isinstance(value, int) or value < 0 for value in board):
        raise EnumerationInvariantError(f"{context}: invalid board; state={key}")
    if any(not isinstance(value, int) or value < 0 for value in captured):
        raise EnumerationInvariantError(f"{context}: invalid captured; state={key}")
    if not state.is_terminal() and player not in (0, 1):
        raise EnumerationInvariantError(f"{context}: invalid current player; state={key}")
    if sum(board) + sum(captured) != TOTAL_SEEDS:
        raise EnumerationInvariantError(
            f"{context}: seed invariant failed; state={key}; "
            f"total={sum(board) + sum(captured)}"
        )


def _write_checkpoint(path: Path, payload: dict[str, Any]) -> None:
    """Write a checkpoint atomically so an interrupted write is not resumable."""
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary = path.with_suffix(path.suffix + ".tmp")
    with temporary.open("wb") as stream:
        pickle.dump(payload, stream, protocol=pickle.HIGHEST_PROTOCOL)
    temporary.replace(path)


def _read_checkpoint(path: Path) -> dict[str, Any]:
    with path.open("rb") as stream:
        payload = pickle.load(stream)
    if payload.get("format_version") != 1:
        raise ValueError(f"Unsupported checkpoint format: {path}")
    return payload


def enumerate_positions(
    max_states: int | None = None,
    progress: bool = True,
    checkpoint_path: Path | None = None,
    resume: bool = False,
) -> dict[str, Any]:
    """Run frontier BFS and return reproducible summary statistics."""
    game = pyspiel.load_game("ayo_olopon")
    if resume:
        if checkpoint_path is None or not checkpoint_path.exists():
            raise FileNotFoundError(f"Checkpoint not found: {checkpoint_path}")
        checkpoint = _read_checkpoint(checkpoint_path)
        frontier = checkpoint["frontier"]
        visited = checkpoint["visited"]
        states_by_depth = Counter(checkpoint["states_by_depth"])
        terminal_states = checkpoint["terminal_states"]
        duplicate_attempts = checkpoint["duplicate_attempts"]
        transitions_generated = checkpoint["transitions_generated"]
        maximum_frontier = checkpoint["maximum_frontier"]
        depth = checkpoint["depth"]
        initial_state = checkpoint["initial_state"]
        if progress:
            print(
                f"Resuming from depth {depth}: frontier={len(frontier):,}, "
                f"visited={len(visited):,}",
                flush=True,
            )
    else:
        initial = game.new_initial_state()
        initial_key = canonical_state(initial)
        validate_state(initial, initial_key, "initial state")
        initial_state = {
            "board": list(initial_key[0]),
            "captured": list(initial_key[1]),
            "current_player": initial_key[2],
        }

        visited = {initial_key}
        frontier = [initial]
        states_by_depth = Counter({0: 1})
        terminal_states = 1 if initial.is_terminal() else 0
        duplicate_attempts = 0
        transitions_generated = 0
        maximum_frontier = 1
        depth = 0
    enumeration_complete = True

    tracemalloc.start()
    started = time.perf_counter()
    while frontier:
        next_frontier = []
        for state_index, state in enumerate(frontier):
            if state.is_terminal():
                continue

            actions = list(state.legal_actions())
            for action in actions:
                transitions_generated += 1
                child = state.child(action)
                child_key = canonical_state(child)
                validate_state(
                    child,
                    child_key,
                    f"depth={depth + 1}, state_index={state_index}, action={action}",
                )
                if child_key in visited:
                    duplicate_attempts += 1
                    continue
                visited.add(child_key)
                next_frontier.append(child)
                states_by_depth[depth + 1] += 1
                if child.is_terminal():
                    terminal_states += 1
                if max_states is not None and len(visited) >= max_states:
                    enumeration_complete = False
                    frontier = []
                    break

            if not enumeration_complete:
                break

        depth += 1
        maximum_frontier = max(maximum_frontier, len(next_frontier))
        frontier = next_frontier if enumeration_complete else []
        if progress:
            print(
                f"BFS depth {depth}: frontier={len(frontier):,}, "
                f"visited={len(visited):,}",
                flush=True,
            )
        if checkpoint_path is not None:
            _write_checkpoint(
                checkpoint_path,
                {
                    "format_version": 1,
                    "frontier": frontier,
                    "visited": visited,
                    "states_by_depth": dict(states_by_depth),
                    "terminal_states": terminal_states,
                    "duplicate_attempts": duplicate_attempts,
                    "transitions_generated": transitions_generated,
                    "maximum_frontier": maximum_frontier,
                    "depth": depth,
                    "initial_state": initial_state,
                },
            )

    elapsed = time.perf_counter() - started
    _, peak_bytes = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    return {
        "search_method": "Frontier-Based BFS",
        "state_representation": "(board, captured, current_player)",
        "initial_state": initial_state,
        "unique_reachable_states": len(visited),
        "terminal_states": terminal_states,
        "states_by_bfs_depth": {str(k): v for k, v in sorted(states_by_depth.items())},
        "maximum_bfs_depth": max(states_by_depth),
        "maximum_frontier_size": maximum_frontier,
        "peak_visited_set_size": len(visited),
        "duplicate_transposition_attempts": duplicate_attempts,
        "total_transitions_generated": transitions_generated,
        "runtime_seconds": elapsed,
        "peak_tracemalloc_bytes": peak_bytes,
        "enumeration_complete": enumeration_complete,
    }


def print_report(summary: dict[str, Any]) -> None:
    """Print a concise human-readable report."""
    print("=== Ayo Olopon Positional State-Space Enumeration ===")
    print(f"Search method: {summary['search_method']}")
    print(f"State representation: {summary['state_representation']}")
    print(f"Initial state: {summary['initial_state']}")
    print(f"Unique reachable states: {summary['unique_reachable_states']:,}")
    print(f"Terminal states: {summary['terminal_states']:,}")
    print(f"Maximum BFS depth: {summary['maximum_bfs_depth']:,}")
    print(f"Maximum frontier size: {summary['maximum_frontier_size']:,}")
    print(f"Duplicate/transposition detections: {summary['duplicate_transposition_attempts']:,}")
    print(f"Total transitions generated: {summary['total_transitions_generated']:,}")
    print(f"Runtime: {summary['runtime_seconds']:.3f} seconds")
    print(f"Peak tracemalloc memory: {summary['peak_tracemalloc_bytes']:,} bytes")
    print(f"Enumeration complete: {'YES' if summary['enumeration_complete'] else 'NO'}")


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--output",
        type=Path,
        default=DEFAULT_OUTPUT,
        help=f"JSON report path (default: {DEFAULT_OUTPUT})",
    )
    parser.add_argument(
        "--max-states",
        type=int,
        default=None,
        help="Optional safety bound; a bounded run is reported as incomplete.",
    )
    parser.add_argument(
        "--quiet",
        action="store_true",
        help="Suppress per-depth progress output.",
    )
    parser.add_argument(
        "--checkpoint",
        type=Path,
        default=None,
        help="Write a resumable pickle checkpoint after each BFS depth.",
    )
    parser.add_argument(
        "--resume",
        action="store_true",
        help="Resume from the checkpoint supplied with --checkpoint.",
    )
    args = parser.parse_args()

    try:
        summary = enumerate_positions(
            max_states=args.max_states,
            progress=not args.quiet,
            checkpoint_path=args.checkpoint,
            resume=args.resume,
        )
    except EnumerationInvariantError as error:
        print(f"INVARIANT FAILURE: {error}", file=sys.stderr)
        return 2

    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(json.dumps(summary, indent=2) + "\n", encoding="utf-8")
    print_report(summary)
    print(f"JSON report: {args.output}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: experiments/enumerate_state_space.cc
#include "Model/ayo_olopon_C++/ayo_olopon.h"

#include <absl/container/flat_hash_set.h>

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

#ifdef _WIN32
#include <windows.h>
#include <psapi.h>
#else
#include <sys/resource.h>
#endif

namespace {

using open_spiel::Action;
using open_spiel::Game;
using open_spiel::State;

struct StateKey {
  uint64_t low = 0;
  uint64_t high = 0;
  bool operator==(const StateKey& other) const {
    return low == other.low && high == other.high;
  }
};

struct StateKeyHash {
  size_t operator()(const StateKey& key) const {
    uint64_t x = key.low ^ (key.high + 0x9e3779b97f4a7c15ULL +
                            (key.low << 6) + (key.low >> 2));
    x ^= x >> 30;
    x *= 0xbf58476d1ce4e5b9ULL;
    x ^= x >> 27;
    x *= 0x94d049bb133111ebULL;
    x ^= x >> 31;
    return static_cast<size_t>(x);
  }
};

StateKey CanonicalState(const State& state) {
  const auto* ayo = dynamic_cast<const open_spiel::ayo_olopon::AyoState*>(&state);
  if (ayo == nullptr) throw std::runtime_error("State is not Ayo Olopon");
  const auto& board = ayo->Board();
  if (board.seeds.size() != 12 || board.score.size() != 2)
    throw std::runtime_error("Unexpected Ayo board dimensions");

  StateKey key;
  int bit = 0;
  auto append = [&](int value, const char* field) {
    if (value < 0 || value > 48)
      throw std::runtime_error(std::string("Invalid ") + field + " value");
    const uint64_t v = static_cast<uint64_t>(value);
    if (bit < 64) {
      key.low |= v << bit;
      if (bit > 58) key.high |= v >> (64 - bit);
    } else {
      key.high |= v << (bit - 64);
    }
    bit += 6;
  };
  for (int value : board.seeds) append(value, "seed");
  for (int value : board.score) append(value, "captured");
  append(ayo->IsTerminal() ? 2 : ayo->CurrentPlayer(), "player");
  return key;
}

void CheckInvariant(const State& state, int depth, Action action) {
  const auto* ayo = dynamic_cast<const open_spiel::ayo_olopon::AyoState*>(&state);
  const auto& board = ayo->Board();
  int total = 0;
  for (int value : board.seeds) total += value;
  for (int value : board.score) total += value;
  if (total != 48) {
    std::cerr << "Invariant failure at depth " << depth << ", action "
              << action << ": total=" << total << "\n" << state << "\n";
    throw std::runtime_error("48-seed conservation invariant failed");
  }
}

uint64_t PeakResidentBytes() {
#ifdef _WIN32
  PROCESS_MEMORY_COUNTERS counters{};
  return GetProcessMemoryInfo(GetCurrentProcess(), &counters, sizeof(counters))
             ? static_cast<uint64_t>(counters.PeakWorkingSetSize)
             : 0;
#else
  struct rusage usage{};
  return getrusage(RUSAGE_SELF, &usage) == 0
             ? static_cast<uint64_t>(usage.ru_maxrss) * 1024ULL
             : 0;
#endif
}

struct Metrics {
  std::vector<uint64_t> states_by_depth;
  uint64_t terminal_states = 0;
  uint64_t duplicate_attempts = 0;
  uint64_t transitions = 0;
  uint64_t peak_frontier = 0;
  bool complete = false;
};

void WriteJson(const std::string& path, const Metrics& m, uint64_t elapsed_ms,
               uint64_t peak_rss) {
  std::ofstream out(path);
  if (!out) throw std::runtime_error("Cannot open output: " + path);
  uint64_t total = 0;
  for (uint64_t n : m.states_by_depth) total += n;
  out << "{\n  \"complete\": " << (m.complete ? "true" : "false")
      << ",\n  \"total_unique_states\": " << total
      << ",\n  \"states_by_depth\": [\n";
  for (size_t i = 0; i < m.states_by_depth.size(); ++i) {
    out << "    {\"depth\": " << i << ", \"states\": "
        << m.states_by_depth[i] << "}"
        << (i + 1 == m.states_by_depth.size() ? "\n" : ",\n");
  }
  out << "  ],\n  \"max_depth\": "
      << (m.states_by_depth.empty() ? 0 : m.states_by_depth.size() - 1)
      << ",\n  \"terminal_states\": " << m.terminal_states
      << ",\n  \"duplicate_attempts\": " << m.duplicate_attempts
      << ",\n  \"legal_transitions_generated\": " << m.transitions
      << ",\n  \"peak_frontier\": " << m.peak_frontier
      << ",\n  \"elapsed_seconds\": " << std::fixed << std::setprecision(3)
      << elapsed_ms / 1000.0 << ",\n  \"peak_rss_bytes\": " << peak_rss
      << "\n}\n";
}

int Run(int argc, char** argv) {
  if (argc > 2) throw std::runtime_error("Usage: enumerate_state_space [output.json]");
  const std::string output = argc == 2 ? argv[1] : "ayo_state_space_results.json";
  const auto start = std::chrono::steady_clock::now();
  const std::shared_ptr<const Game> game = open_spiel::LoadGame("ayo_olopon");
  std::vector<std::unique_ptr<State>> frontier;
  frontier.push_back(game->NewInitialState());

  absl::flat_hash_set<StateKey, StateKeyHash> visited;
  Metrics m;
  CheckInvariant(*frontier.front(), 0, -1);
  visited.insert(CanonicalState(*frontier.front()));
  m.states_by_depth.push_back(1);
  m.peak_frontier = 1;

  for (int depth = 0; !frontier.empty(); ++depth) {
    std::vector<std::unique_ptr<State>> next;
    for (const auto& state : frontier) {
      if (state->IsTerminal()) {
        ++m.terminal_states;
        continue;
      }
      for (Action action : state->LegalActions()) {
        ++m.transitions;
        std::unique_ptr<State> child = state->Child(action);
        CheckInvariant(*child, depth + 1, action);
        if (!visited.insert(CanonicalState(*child)).second) {
          ++m.duplicate_attempts;
        } else {
          next.push_back(std::move(child));
        }
      }
    }
    if (next.empty()) break;
    m.states_by_depth.push_back(next.size());
    m.peak_frontier = std::max<uint64_t>(m.peak_frontier, next.size());
    std::cout << "BFS depth " << depth + 1 << ": frontier=" << next.size()
              << ", visited=" << visited.size() << "\n" << std::flush;
    frontier = std::move(next);
  }

  m.complete = true;
  const auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
      std::chrono::steady_clock::now() - start);
  const uint64_t peak_rss = PeakResidentBytes();
  WriteJson(output, m, elapsed.count(), peak_rss);
  uint64_t total = 0;
  for (uint64_t n : m.states_by_depth) total += n;
  std::cout << "Enumeration complete: YES\n"
            << "Total unique positional states: " << total << "\n"
            << "Maximum BFS depth: " << m.states_by_depth.size() - 1 << "\n"
            << "Terminal states: " << m.terminal_states << "\n"
            << "Duplicate attempts: " << m.duplicate_attempts << "\n"
            << "Legal transitions: " << m.transitions << "\n"
            << "Peak RSS: " << peak_rss << " bytes\n"
            << "Results: " << output << "\n";
  return 0;
}

}  // namespace

int main(int argc, char** argv) {
  try {
    return Run(argc, argv);
  } catch (const std::exception& error) {
    std::cerr << "Enumeration failed: " << error.what() << "\n";
    return 1;
  }
}
